### PR TITLE
Add proper CPU lowering for 2D `tosa::CustomOp` with arbitrary stride, dilation, and input padding

### PR DESCRIPTION
**Pull request by justinrosner**
## Motivation
This PR implements proper CPU lowering for `tosa::customOp` operations (representing 2D transpose convolutions) by adding support for dilation and input padding attributes. This change means that we can now remove the LIT test workaround in the bwd_data_conv e2e tests that were introduced in: https://github.com/ROCm/rocMLIR/pull/1951

Implements: https://github.com/ROCm/rocMLIR-internal/issues/1990

## Technical Details
This PR implements the following key changes:
- Update `TransposeConvNonStridedConverter` and `TransposeConvStridedConverter` to both handle input padding and dilation attributes (see the comments in the new logic for more details on this)
- Updated the bwd_data_conv e2e tests that no longer need a workaround.
- Now that we will no longer be using upstream Tosa for CPU/GPU lowering, I've removed any of the additional changes that were added in the initial work to support lowering from MIGraphX

## Test Plan
- Update existing bwd_conv e2e tests to use the new lowering path
- Add new e2e tests for more edge cases

## Test Result
- [ ] PR CI (with MIGraphX tests enabled)
- [x] All bwd_conv e2e LIT tests are passing

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

**Comment by justinrosner** [comment]
Note for reviewers, once I've gotten the necessary approvals I will separate this into two different commits. One for the external llvm-project changes, and another for the rocMLIR changes.

File: external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaDecomposeTransposeConv.cpp [hunk 1]
```diff
@@ -105,9 +143,41 @@ class TransposeConvStridedConverter
...
+          wShape[1],
+          wShape[2],
+          wShape[0]  // K becomes I
+      };
+      auto swappedTy =
+          RankedTensorType::get(swappedShape, weightTy.getElementType());
+      weight = rewriter.create<tosa::TransposeOp>(
+          loc, swappedTy, weight, rewriter.getDenseI32ArrayAttr({3, 1, 2, 0}));
+      weightTy = cast<ShapedType>(weight.getType());
+    }
```
Comment (justinrosner) [inline external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaDecomposeTransposeConv.cpp]: Would it make more sense to do this in MIGraphXToTosa, and then again when converting back from TosaToRock?

**Comment by dhernandez0** [comment]
do we want to have an upstream PR for this? they'll probably say it's not part of the tosa specs, but maybe we can open a discussion to include this?

**Comment by justinrosner** [comment]
> do we want to have an upstream PR for this? they'll probably say it's not part of the tosa specs, but maybe we can open a discussion to include this?

They explicitly had PRs removing dilation attributes from the code a couple of years ago. It still may be worth a try. Do you know what forum would be best to go about asking for changes to the TOSA spec?

**Comment by pabloantoniom** [comment]
In `MIGraphXToTosa` we do:

```c++
// Set attributes common to both forwards and backwards conv
cop->setAttr("dilation", rewriter.getDenseI64ArrayAttr(dilations));
cop->setAttr("stride", rewriter.getDenseI64ArrayAttr(strides));
cop->setAttr("acc_type", TypeAttr::get(accType));
```

But I don't see anywhere in this PR where those attributes are used. Is that intentional?

**Comment by justinrosner** [comment]
> In `MIGraphXToTosa` we do:
> 
> ```c++
> // Set attributes common to both forwards and backwards conv
> cop->setAttr("dilation", rewriter.getDenseI64ArrayAttr(dilations));
> cop->setAttr("stride", rewriter.getDenseI64ArrayAttr(strides));
> cop->setAttr("acc_type", TypeAttr::get(accType));
> ```
> 
> But I don't see anywhere in this PR where those attributes are used. Is that intentional?

They are both used in `TosaDecomposeTransposeConv`. For the simple, non-strided case, we pass dilation through and use the values directly on the forwards conv op that is created. And for the strided case we use both the strides and dilation in the calculation. Also `acc_type` is being set for both.

**Comment by justinrosner** [comment]
@pabloantoniom This has now been rebased to make changes to RocmlirCustomTosaDecompose instead of changes directly to external/llvm-project

File: external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaDecomposeTransposeConv.cpp [hunk 2]
```diff
@@ -105,10 +105,6 @@ class TransposeConvStridedConverter
     llvm::ArrayRef<int64_t> pad = op.getOutPad();
     llvm::ArrayRef<int64_t> stride = op.getStride();
 
-    // If striding is all 1 we can modify padding and reverse the kernel along
```
Comment (pabloantoniom) [inline external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaDecomposeTransposeConv.cpp:108]: I think you want to keep this comment to make it consistent with upstream

File: mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp [hunk 3]
```diff
@@ -72,12 +100,28 @@ class TransposeConvNonStridedConverter
     ShapedType resultTy = cast<ShapedType>(op->getResult(0).getType());
 
     // Translate acc_type, padding and stride attributes.
-    llvm::ArrayRef<int64_t> pad = cast<DenseI64ArrayAttr>(op->getAttr("pad"));
+    llvm::ArrayRef<int64_t> pad = cast<DenseI64ArrayAttr>(op->getAttr("out_pad"));
```
Comment (pabloantoniom) [inline mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp]: I think you will need to adapt `rocmlir-custom-tosa-decompose.mlir` to this change. I don't think it will pass after this code change

File: mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp [hunk 4]
```diff
@@ -92,14 +136,43 @@ class TransposeConvNonStridedConverter
...
+    // convPadLeft   = effKWm1 - inPadLeft   + outPadLeft
+    // convPadRight  = effKWm1 - inPadRight  + outPadRight
+    SmallVector<int64_t, 4> convPad = {
+        effKHm1 - inPadVals[0] + pad[0],
+        effKHm1 - inPadVals[1] + pad[1],
+        effKWm1 - inPadVals[2] + pad[2],
+        effKWm1 - inPadVals[3] + pad[3]
+    };
+
+    bool needSlice = false;
```
Comment (pabloantoniom) [inline mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp:340]: What does needSlice signify? Can you add a comment?

File: mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp [hunk 5]
```diff
@@ -360,11 +556,51 @@ class TransposeConvStridedConverter : public OpRewritePattern<tosa::CustomOp> {
...
+    if (kWPrime == 1 && lostW > 0)
+      effPadLeft += lostW / 2;
+
+    int64_t resultSliceTop;
+    int64_t resultSliceLeft;
+    int64_t resultPadTop;
+    int64_t resultPadLeft;
+    // Convert effective padding into slice (crop) and post-pad just like the
+    // prior logic but now using effPad*.
+    if (op->hasAttr("pad")) {
```
Comment (pabloantoniom) [inline mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp:777]: Shouldn't it be the other way around? If pad attr is available then access `pad`?

File: external/llvm-project/mlir/lib/Dialect/Tosa/IR/TosaOps.cpp [hunk 6]
```diff
@@ -3630,80 +3630,37 @@ LogicalResult TransposeConv2DOp::verify() {
...
-        if (outputHeight != (inputHeight - 1) * strideY + outPadTop +
-                                outPadBottom + kernelHeight)
-          return emitOpError(
-                     "dimension mismatch: expected OH == (IH - 1) * stride_y "
-                     "+ out_pad_top + out_pad_bottom + KH, but got ")
-                 << outputHeight << " != (" << inputHeight << " - 1) * "
-                 << strideY << " + " << outPadTop << " + " << outPadBottom
-                 << " + " << kernelHeight;
-      }
+    if (ShapedType::isStatic(inputHeight) &&
```
Comment (pabloantoniom) [inline external/llvm-project/mlir/lib/Dialect/Tosa/IR/TosaOps.cpp:3639]: It's great we can get rid of changes to `external`. This should have been made in my PR https://github.com/ROCm/rocMLIR/pull/2016, so thanks for the changes. At the same time, it makes me think that our tosa::CustomOp is now weaker as it's not checking for all this. I really dislike hacking tosa.

File: mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp [hunk 7]
```diff
@@ -72,12 +100,28 @@ class TransposeConvNonStridedConverter
...
     // Translate acc_type, padding and stride attributes.
-    llvm::ArrayRef<int64_t> pad = cast<DenseI64ArrayAttr>(op->getAttr("pad"));
+    llvm::ArrayRef<int64_t> pad = cast<DenseI64ArrayAttr>(op->getAttr("out_pad"));
     llvm::ArrayRef<int64_t> stride =
         cast<DenseI64ArrayAttr>(op->getAttr("stride"));
     auto accTypeAttr = cast<TypeAttr>(op->getAttr("acc_type"));
     Type accType = accTypeAttr.getValue();
 
+    // Fetch dilation (default {1,1})
+    SmallVector<int64_t, 2> dilationVals = {1, 1};
```
Comment (pabloantoniom) [inline mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp]: I don't like that this PR is all hardcoded for Conv2D. It's fair since the that's the scope of the PR, but at the same time I wonder how difficult will it be to make this work for Conv3D which is coming very soon.

File: external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaDecomposeTransposeConv.cpp (hunk 2 above)
Comment (justinrosner) [reply external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaDecomposeTransposeConv.cpp:108]: Yep, meant to add that back!

File: mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp (hunk 3 above)
Comment (justinrosner) [reply mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp]: I updated the LIT test but forgot to push the changes. It should be updated now.

File: mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp (hunk 5 above)
Comment (justinrosner) [reply mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp:777]: This was a little confusing because in the original Tosa file they named the output padding `pad`. I just fixed that to `outPad`, which should make logic like this cleaner.

Essentially this piece of logic is checking if the op has an input padding attribute in which case it will use the effective padding that it just computed, and if it doesn't then it falls back to just using the output pad as part of the calculation.

File: external/llvm-project/mlir/lib/Dialect/Tosa/IR/TosaOps.cpp (hunk 6 above)
Comment (justinrosner) [reply external/llvm-project/mlir/lib/Dialect/Tosa/IR/TosaOps.cpp:3639]: What do you think about incorporating some of that lost verification logic at the start of RocmlirCustomTosaDecompose? If the CustomOp matches the pattern, then we run the verification checks before we do any of the transformations?

File: mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp (hunk 7 above)
Comment (justinrosner) [reply mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp]: I suspect that the 3D case will be a little tricky. I've opened up a case to track the work that is needed for that:
https://github.com/ROCm/rocMLIR-internal/issues/2061

File: external/llvm-project/mlir/lib/Dialect/Tosa/IR/TosaOps.cpp (hunk 6 above)
Comment (pabloantoniom) [reply external/llvm-project/mlir/lib/Dialect/Tosa/IR/TosaOps.cpp:3639]: That's actually a great idea, I like it!

**Comment by pabloantoniom** [comment]
It seems to be failing on CI due to tosa::CustomOp lowering issues, LMK if you need some help with that

**Comment by justinrosner** [comment]
> It seems to be failing on CI due to tosa::CustomOp lowering issues, LMK if you need some help with that

This was related to the fact that this pass wasn't running in the highlevel pipeline. When this PR turned that on,RocmlirCustomTosaDecompose would start throwing an error because there were still tosa::CustomOps that were yet to be lowered. This just needed an update the the legality check.

File: external/llvm-project/mlir/lib/Dialect/Tosa/IR/TosaOps.cpp (hunk 6 above)
Comment (justinrosner) [reply external/llvm-project/mlir/lib/Dialect/Tosa/IR/TosaOps.cpp:3639]: Updated!

File: mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp [hunk 8]
```diff
@@ -45,6 +45,184 @@ struct RocmlirCustomTosaDecomposePass
   void runOnOperation() override;
 };
 
+// This is mostly a copy of the verification op that exists for
+// tosa::TransposeConv2DOp in upstream with the output shape checks updated
+// to properly account for input padding and dilation.
+// See here for the formula being used:
+// https://onnx.ai/onnx/operators/onnx__ConvTranspose.html
+LogicalResult verifyCustomOp(tosa::CustomOp op) {
```
Comment (dhernandez0) [inline mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp]: verifyConvTranspose?

File: mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp [hunk 9]
```diff
@@ -45,6 +45,184 @@ struct RocmlirCustomTosaDecomposePass
...
+  }
+
+  // Rest of the checks depend on the output type being a RankedTensorType
+  const auto outputType =
+      llvm::dyn_cast<RankedTensorType>(op.getResult(0).getType());
+  if (!outputType)
+    return success();
+
+  // Fetch pad & dilation;
+  SmallVector<int64_t, 2> dilation = {1, 1};
```
Comment (dhernandez0) [inline mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp:134]: I think all the check must be adapted to work for 3D convs as well. Ideally for any ND conv.

File: mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp [hunk 10]
```diff
@@ -45,6 +45,184 @@ struct RocmlirCustomTosaDecomposePass
...
+      biasChannels != outputChannels && biasChannels != 1)
+    return op.emitOpError(
+               "bias channels expected to be equal to output channels (")
+           << outputChannels << ") or 1, got " << biasChannels;
+
+  return success();
+}
+
+// If this is a backward-data (transpose) conv lowered from MIGraphX, its
+// filter logical in/out channels are reversed relative to forward Conv2D.
```
Comment (dhernandez0) [inline mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp:216]: Conv2D/Conv3D

File: mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp [hunk 11]
```diff
@@ -45,6 +45,184 @@ struct RocmlirCustomTosaDecomposePass
...
+FailureOr<std::tuple<Value, ShapedType>>
+swapInputOutputDimensions(OpBuilder &rewriter, tosa::CustomOp op, Value weight,
+                          ShapedType weightTy) {
+  if (op.getOperatorName() == ROCK_CUSTOMOP_CONV_BWD_DATA) {
+    // Expected current shape: [K, H, W, C] but Conv2D expects [O, H, W, I]
```
Comment (dhernandez0) [inline mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp:221]: what about 3D convs

File: mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp [hunk 12]
```diff
@@ -72,12 +253,29 @@ class TransposeConvNonStridedConverter
...
+    llvm::ArrayRef<int64_t> outPad =
+        cast<DenseI64ArrayAttr>(op->getAttr("out_pad"));
     llvm::ArrayRef<int64_t> stride =
         cast<DenseI64ArrayAttr>(op->getAttr("stride"));
     auto accTypeAttr = cast<TypeAttr>(op->getAttr("acc_type"));
     Type accType = accTypeAttr.getValue();
 
+    // Fetch dilation (default {1,1})
+    SmallVector<int64_t, 2> dilationVals = {1, 1};
+    if (auto dilOpt = cast<DenseI64ArrayAttr>(op->getAttr("dilation"))) {
```
Comment (dhernandez0) [inline mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp:275]: this doesn't seem to work for 3D convs?

File: mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp [hunk 13]
```diff
@@ -61,6 +239,9 @@ class TransposeConvNonStridedConverter
     if (op.getNumResults() != 1)
       return rewriter.notifyMatchFailure(op, "should have 1 result");
 
+    if (failed(verifyCustomOp(op)))
```
Comment (pabloantoniom) [inline mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp]: This is great, but would it make sense to move all the checks above (Make sure this is a transpose conv op) to `verifyCustomOp` as the initial checks? Since the `verifyCustomOp` is intended to check custom ops representing transpose convs.

**Review by pabloantoniom — APPROVED** [review]
LGTM, I just left a minor suggestion

File: mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp (hunk 9 above)
Comment (justinrosner) [reply mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp:134]: I spoke with Daniel offline, all of the work needed to fully support 3D convs is going to be done in https://github.com/ROCm/rocMLIR-internal/issues/2061, so for now we just want to error out if we detect a 3D conv.

File: mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp (hunk 13 above)
Comment (justinrosner) [reply mlir/lib/Conversion/RocmlirCustomTosaDecompose/RocmlirCustomTosaDecompose.cpp]: Makes sense. I moved them in!

**Comment by dhernandez0** [comment]
The title and description of the PR need to be updated, we don't use tosa::transpose_conv2d anymore, right?

Also, I think it's using code from https://github.com/ROCm/rocMLIR/pull/2034 ? can you point to that PR's branch (instead of develop) if that's the case, so it's easier to review.

**Comment by justinrosner** [comment]
> The title and description of the PR need to be updated, we don't use tosa::transpose_conv2d anymore, right?
> 
> Also, I think it's using code from #2034 ? can you point to that PR's branch (instead of develop) if that's the case, so it's easier to review.

I updated the description and title to match the changes to using CustomOp. This change isn't using anything in that branch, just the 2D work that was already checked into develop

File: mlir/lib/Dialect/Rock/Pipelines/Pipelines.cpp [hunk 14]
```diff
@@ -69,6 +69,7 @@ void rock::buildBufferizePipeline(OpPassManager &pm,
     funcPm.addPass(rock::createRockViewToTransformPass());
   }
 
+  funcPm.addPass(createRocmlirCustomTosaDecomposePass());
```
Comment (dhernandez0) [inline mlir/lib/Dialect/Rock/Pipelines/Pipelines.cpp:72]: the pass already existed, where was it used?

**Review by dhernandez0 — APPROVED** [review]

File: mlir/lib/Dialect/Rock/Pipelines/Pipelines.cpp (hunk 14 above)
Comment (justinrosner) [reply mlir/lib/Dialect/Rock/Pipelines/Pipelines.cpp:72]: To my knowledge it wasn't being used anywhere. It was just manually being called in some of the LIT tests, but it wasn't officially part of any pipeline since without this change it wasn't fully functional.

Comment (pabloantoniom) [reply mlir/lib/Dialect/Rock/Pipelines/Pipelines.cpp:72]: Yes, it was added in https://github.com/ROCm/rocMLIR/pull/2016 but because the CPU lowering was broken it was not added to any pipeline (it was exercised on LIT tests though, like Justin mentioned). Now that CPU lowering works it makes sense to have it here
